src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -859,18 +882,25 @@ initSchema = do
       logInfo_ logger $ "initSchema: "  <> fromUtf8 tablename
       callDb "initSchema" $ createVersionedTable tablename
 
-getEndingTxId :: ChainwebVersion -> ChainId -> BlockHeight -> BlockHandler logger SQLiteEnv TxId
-getEndingTxId v cid bh = callDb "getEndingTxId" $ \db -> do
-    if bh == genesisHeight v cid
-      then return 0
-      else
-        qry db "SELECT endingtxid FROM BlockHistory where blockheight = ?"
-            [SInt (fromIntegral $ pred bh)]
-            [RInt]
-          >>= fmap convertInt . expectSingleRowCol "endingtxid for block"
-  where
-    convertInt (SInt thing) = fromIntegral thing
-    convertInt _ = error "impossible"
+getEndTxId :: Text -> ParentContext -> BlockHandler logger SQLiteEnv TxId
+getEndTxId msg pc = case pc of
+  GenesisParentContext {} -> return 0
+  BlockParentContext {} -> getEndTxId' msg (parentContextHeight pc) (parentContextHash pc)
+
+getEndTxId' :: Text -> BlockHeight -> BlockHash -> BlockHandler logger SQLiteEnv TxId
+getEndTxId' msg bh bhsh =
+  callDb "getEndTxId" $ \db -> do
+    r <- qry db
+      "SELECT endingtxid FROM BlockHistory WHERE blockheight = ? and hash = ?;"
+      [ SInt $ fromIntegral bh
+      , SBlob $ runPutS (encodeBlockHash bhsh)
+      ]
+      [RInt]
+    case r of
+      [[SInt tid]] -> return (TxId (fromIntegral tid))
+      [] -> throwM $ BlockHeaderLookupFailure $ msg <> ".getEndTxId: not in db: " <>
+            sshow (bh, bhsh)
+      _ -> internalError $ msg <> ".getEndTxId: expected single-row int result, got " <> sshow r
```

Comment:
I don't think is unsafe per se, it'll just fail if you mismatch the hash and height. It needs to be exported because `getBlockHistory` doesn't have a block header for the block it wants the end txid for, just the hash and height.